[CoreML EP] Implement Unary & Reduce operators

ShukantPal opened the pull request:
### Description

This change is a follow-up to #15327. It adds Unary operators (Sqrt, Reciprocal) and Reduce operators (ReduceSum, ReduceMean). I've tried to follow existing patterns in the code :-)


### Motivation and Context

This reduces fragmentation across EPs when using CoreML on macOS, thereby speeding up execution.

edgchen1 commented on onnxruntime/core/providers/coreml/builders/impl/reduction_op_builder.cc:
```diff
@@ -0,0 +1,71 @@
+
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(model_builder, node);
+
+  if (op_type == "ReduceSum") {
```
there is an attribute 'noop_with_empty_axes' in the ONNX spec:
https://github.com/onnx/onnx/blob/main/docs/Operators.md#reducesum

we should either handle it or not support it (by overriding OpBuilder::IsOpSupported to check for the cases we do support). seems like it can map to the CoreML layer's reduceAll field.

edgchen1 commented on onnxruntime/core/providers/coreml/builders/impl/reduction_op_builder.cc:
```diff
@@ -0,0 +1,71 @@
+  const auto& input_defs(node.InputDefs());
+
+  NodeAttrHelper helper(node);
+  auto axes = helper.Get("axes", std::vector<int64_t>{});
```
in ONNX opset 18, 'axes' is an input instead of an attribute. this changed earlier in opset 13 for ReduceSum.
could possibly first check for the presence of the second, optional, 'axes' input and ensure that it is constant (a constant initializer). if it doesn't exist, then fall back to checking for the attribute.
that probably should be in an override of IsOpSupported(), as a non-constant axes input would not be supported.

edgchen1 commented on onnxruntime/core/providers/coreml/builders/impl/unary_op_builder.cc:
```diff
@@ -0,0 +1,58 @@
+class UnaryOpBuilder : public BaseOpBuilder {
+ private:
+#ifdef __APPLE__
+  [[nodiscard]] Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
```
```suggestion
  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
```

edgchen1 commented on onnxruntime/core/providers/coreml/builders/op_builder_factory.cc:
```diff
@@ -99,6 +99,16 @@ static OpBuilderRegistrations CreateOpBuilderRegistrations() {
   }
 
+  {  // Unary
+    CreateUnaryOpBuilder("Sqrt", op_registrations);
```
please also update tools/ci_build/github/apple/coreml_supported_ops.md

edgchen1 reviewed (commented):
Thanks for your contribution!

ShukantPal commented:
Thanks for your review @edgchen1 ! I've followed up on all of your comments/suggestions. Please let me know how it looks now.

edgchen1 commented:
/azp run MacOS CI Pipeline

natke commented:
Hi @ShukantPal, thank you for this contribution! We're interested in learning more about your use case. Can you tell us a little bit about how you are using ONNX Runtime? Or if you would prefer you can reach me at nakersha@microsoft.com

ShukantPal commented:
Hi @natke, I'm developing a goofy macOS virtual camera that uses different video filters like FaceMesh, CenterFace, DFL, etc. To get real-time frame rates, executing on CoreML / ANE is necessary on M1 MacBooks. I wanted to keep my application cross-platform to run on Intel Macs (with discrete GPUs) and Windows in the future. That's why I chose ONNX runtime, but having full CoreML support is still necessary for satisfactory performance on M1.

natke commented:
Sounds pretty awesome. Is it published anywhere? I'd love to see a demo

ShukantPal commented:
@natke Haha, not yet − I've been working on it. Can send you a beta build when ready :-)

ShukantPal commented on onnxruntime/core/providers/coreml/builders/impl/reduction_op_builder.cc:
```diff
@@ -0,0 +1,115 @@
+
+  NodeAttrHelper helper(node);
+  if (input_defs.size() > 1) {
+    auto& axes_tensor = *initializers.at(input_defs[1]->Name());
+    const int64_t* raw_axes = axes_tensor.int64_data().empty()
+      ? reinterpret_cast<const int64_t*>(axes_tensor.raw_data().data())
+      : axes_tensor.int64_data().data();
+    const auto size = axes_tensor.dims()[0];
```
Not sure how that'd work. DataAsSpan returns a gsl::span over a std::vector.

Is that the preferred container to use? Since the Initializer is in an if scope, it would be destroyed before the axes data is used so need to keep that in mind as well.

edgchen1 replied:
sorry, to clarify, was suggesting to use the helper Initializer type instead of dealing with the TensorProto details. you can still copy from the span into the axes vector.

edgchen1 replied:
generally, gsl::span and InlinedVector use is preferred as appropriate.
https://github.com/microsoft/onnxruntime/blob/9db24f8fec40c047044094441c90bedfe05bae6f/docs/Coding_Conventions_and_Standards.md?plain=1#L15
https://github.com/microsoft/onnxruntime/blob/9db24f8fec40c047044094441c90bedfe05bae6f/docs/Coding_Conventions_and_Standards.md?plain=1#L71

however, when the CoreML EP was initially written those guidelines may not have been in place yet. using std::vector is fine and consistent here.

ShukantPal replied:
done!

skottmckay commented on tools/ci_build/github/apple/coreml_supported_ops.md:
```diff
@@ -1,34 +1,38 @@
-|ai.onnx:Tanh||
-|ai.onnx:Transpose||
+| Operator                   | Note                                                                                                                                                      |
+|----------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------|
+| ai.onnx:Add                ||
```
This formatting change would break [this](https://github.com/microsoft/onnxruntime/blob/9db24f8fec40c047044094441c90bedfe05bae6f/tools/python/util/mobile_helpers/usability_checker.py#L35-L48) tool

ShukantPal replied:
Thanks for the heads up; IntelliJ was trying to be a smart alec 🤦. I've reverted the formatting & the diff is much cleaner now.

edgchen1 commented on onnxruntime/core/providers/coreml/builders/impl/reduction_op_builder.cc:
```diff
@@ -0,0 +1,118 @@
+                                           const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+
+  if (input_defs.size() > 1) {
```
```suggestion
  if (input_defs.size() > 1 && input_defs[1]->Exists()) {
```

edgchen1 commented on onnxruntime/core/providers/coreml/builders/impl/reduction_op_builder.cc:
```diff
@@ -0,0 +1,118 @@
+                         const logging::Logger& logger) const override;
+};
+
+namespace {
+template <typename T>
+void AddReductionParams(T* params, const std::vector<int64_t>& axes, bool keepdims, bool noop_with_empty_axes) {
+  params->set_keepdims(keepdims);
+
+  for (auto& axis : axes)
+    params->add_axes(axis);
+
+  if (axes.size() == 0 && !noop_with_empty_axes)
+    params->set_reduceall(true);
+}
+} // namespace
```
nit: this can go in the `#ifdef __APPLE__` too

ShukantPal replied:
Done!

edgchen1 commented:
/azp run MacOS CI Pipeline

ShukantPal commented:
git is hard 🤦 . Let me see if I can fix this

ShukantPal commented on onnxruntime/core/providers/coreml/builders/impl/reduction_op_builder.cc:
```diff
@@ -0,0 +1,119 @@
+    int64_t* data = axes_initializer.data<int64_t>();
+    int64_t size = axes_initializer.size();
+
+    axes = std::vector<int64_t>(data, data + size);
```
@edgchen1 I changed this code to use Initializer.

ShukantPal commented on onnxruntime/core/providers/coreml/builders/helper.cc:
```diff
@@ -35,6 +35,11 @@ bool GetShape(const NodeArg& node_arg, std::vector<int64_t>& shape, const loggin
 }
 
 bool IsNodeSupported(const Node& node, const GraphViewer& graph_viewer, const logging::Logger& logger) {
+  if (node.Name() == "Square_1") {
+    LOGS(logger, WARNING) << "CoreML skipping Square_1";
+    return false;
+  }
+
```
```suggestion
  if (node.Name() == "Square_1") {
    LOGS(logger, WARNING) << "CoreML skipping Square_1";
    return false;
  }

```

ShukantPal replied:
```suggestion
```

ShukantPal commented on onnxruntime/core/providers/coreml/builders/helper.cc:
```diff
@@ -35,6 +35,11 @@ bool GetShape(const NodeArg& node_arg, std::vector<int64_t>& shape, const loggin
 }
 
 bool IsNodeSupported(const Node& node, const GraphViewer& graph_viewer, const logging::Logger& logger) {
+  if (node.Name() == "Square_1") {
```
My bad; I thought that was removed in a previous commit.

Removed it.

edgchen1 commented:
/azp run MacOS CI Pipeline

skottmckay commented on onnxruntime/core/providers/coreml/builders/impl/reduction_op_builder.cc:
```diff
@@ -0,0 +1,119 @@
+    const auto& initializers = input_params.graph_viewer.GetAllInitializedTensors();
+    if (!Contains(initializers, axes_name)) {
+      LOGS(logger, VERBOSE) << "Axes of reduction must be a constant initializer";
+      return false;
```
The changes look fine to me but I think there's a small testing gap. The existing unit tests for ReduceMean don't seem to have a test that adds `axes` as a input using an initializer. ReduceSum has a couple. e.g. [https://github.com/microsoft/onnxruntime/blob/b20d5e85d5ca2333e9c0f52b595254c41c40f7ab/onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc#L2363](https://github.com/microsoft/onnxruntime/blob/b20d5e85d5ca2333e9c0f52b595254c41c40f7ab/onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc#L2363)

Could you please add a test to ensure we have unit test coverage of the ReduceSum path for recent opsets where `axes` is an input? The ReduceSum test is a good example 
- use AddInput with the `is_initializer` parameter set to `true` 
  - vs. older tests using AddAttribute
- set opset to 18 in the OpTester constructor 
  - axes was moved to an input for ReduceSum in opset 13 and the others in opset 18

Can probably cut-and-paste an existing test or two and just tweak it with those changes.

ShukantPal replied:
Thanks @skottmckay. I implemented 2 tests for ReduceMean op with axes as an initializer input.

skottmckay commented:
/azp run Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,MacOS CI Pipeline

skottmckay commented:
/azp run orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-python-checks-ci-pipeline,onnxruntime-binary-size-checks-ci-pipeline

skottmckay commented:
Could you please run the linting tools as well? 

https://github.com/microsoft/onnxruntime/actions/runs/4954869911/jobs/8903556636?pr=15532

Instructions for installing locally if you don't want to manually address the issues: https://github.com/microsoft/onnxruntime/blob/main/docs/Coding_Conventions_and_Standards.md#linting

ShukantPal commented:
@skottmckay I have

* added 2 unit tests for ReduceMean with axes input
* fixed issues in 2 existing unit tests in ReduceSum after CoreML. I had to suppress one test when using CoreML EP because expected loss of precision on CoreML.
* fixed linter issues manually (lintrunner doesn't work for me on my macOS workstation. "error:        No such file or directory (os error 2)")

Let me know if you find any miscellaneous issues in code review!

skottmckay commented:
> * fixed linter issues manually (lintrunner doesn't work for me on my macOS workstation. "error:        No such file or directory (os error 2)")

Looks to be a couple of (painfully) minor clang-format errors to address.

@justinchuby is there a way to fix the macos issue? I see the same issue:

```
> pip3 install lintrunner lintrunner-adapters                                                                                                                                                    
Defaulting to user installation because normal site-packages is not writeable
Collecting lintrunner
  Downloading lintrunner-0.10.7-py3-none-macosx_10_9_x86_64.macosx_11_0_arm64.macosx_10_9_universal2.whl (3.0 MB)
     |████████████████████████████████| 3.0 MB 21.8 MB/s
Collecting lintrunner-adapters
  Downloading lintrunner_adapters-0.8.0-py3-none-any.whl (55 kB)
     |████████████████████████████████| 55 kB 13.0 MB/s
Requirement already satisfied: click<9.0.0,>=8.1.3 in /Users/me/Library/Python/3.9/lib/python/site-packages (from lintrunner-adapters) (8.1.3)
Installing collected packages: lintrunner-adapters, lintrunner
Successfully installed lintrunner-0.10.7 lintrunner-adapters-0.8.0
WARNING: You are using pip version 21.2.4; however, version 23.1.2 is available.
You should consider upgrading via the '/Applications/Xcode.app/Contents/Developer/usr/bin/python3 -m pip install --upgrade pip' command.

> lintrunner init
[2023-05-18T22:00:51Z INFO lintrunner::linter] Initializing linter: 'RUFF'
error:        No such file or directory (os error 2)

> whereis lintrunner                                                                                                                                                                             
lintrunner: /Users/me/Library/Python/3.9/bin/lintrunner
```

skottmckay commented:
/azp run Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,MacOS CI Pipeline

skottmckay commented:
/azp run orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-python-checks-ci-pipeline,onnxruntime-binary-size-checks-ci-pipeline

ShukantPal commented:
* I ran the linter on Windows and fixed the lining issues.
* It seems like TensorRT, OpenVINO, and DNN don't support ReduceMean opset 18 with axes input, so I disabled those providers in the new tests (copying that from similar tests for ReduceMean).

But I need help understanding why the test_layer_normalization* tests are failing on CoreML/macOS. The ReduceMean op-tests pass, but still there's some mismatch with downstream "Reshape" layers in the test models.

I don't have this issue when testing the ReduceMean layers in my own models.

ShukantPal commented on tools/ci_build/github/apple/coreml_supported_ops.md:
```diff
@@ -25,11 +25,15 @@ Keep in sync with doco generated from /docs/execution-providers/CoreML-Execution
 |ai.onnx:Pow|Only supports cases when both inputs are fp32.|
 |ai.onnx:PRelu|Input slope should be constant.<br/>Input slope should either have shape [C, 1, 1] or have 1 element.|
+|ai.onnx:Reciprocal||
+|ai.onnx.ReduceMean||     
```
```suggestion
```

ShukantPal commented:
@skottmckay 

Given the testing complications, I've commented out the line registering ReduceMean. Hopefully, rest of the PR provides enough value for it to be merged :-)

skottmckay commented:
/azp run orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-python-checks-ci-pipeline,onnxruntime-binary-size-checks-ci-pipeline

ShukantPal commented:
@skottmckay Here were the failing tests: https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1012534&view=logs&j=07136875-e4b3-5a15-ff50-026d13f57d01&t=b16e8f5d-fb27-5dd3-278d-cb1fb09fb81a

They come from the --build_wheel option.

skottmckay commented:
Definitely want to get it added, but I think we need to figure out the root cause given we're calling AddReductionParams the same way for both ReduceMean and ReduceSum, so if there's an issue I would expect it potentially applies to both operator types and it's possibly that there's no test showing the issue for ReduceSum.

The failing tests are ONNX test cases. We create a binary called onnx_test_runner that can be used to execute/debug. The ONNX tests are in ./cmake/external/onnx/onnx/backend/test/data/node. Specify the EP with '-e'. CPU passes, CoreML fails.

e.g. `./build/MacOS/Debug/onnx_test_runner -e coreml ./cmake/external/onnx/onnx/backend/test/data/node/test_layer_normalization_2d_axis0_expanded`

I don't think this is the root cause, but the last param for the CoreML parms is reduceAll which might not be 1:1 with noop_with_empty_axes. 

https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#reducemeanlayerparams

CoreML reduceAll says to ignore the axes parameter and reduce all. 

So if axes is empty, reduceAll is true, but if noop_with_empty_axes is also true the ONNX spec says do nothing which doesn't seem to have an equivalent in CoreML. In that case we could probably drop the node in CoreML or insert an Identity node to do the value rename given it's turned into a no-op.

skottmckay commented:
Sorry - I overlooked the early exit for noop_with_empty_axes. It would be good to clarify the implementation in ReductionOpBuilder::AddToModelBuilderImpl though and set a reduceAll value instead of passing noop_with_empty_axes into the AddReductionParms call.

skottmckay commented:
Might be an issue with Flatten. There's not a lot of places in this particular test where this combination of shapes is possible to get wrong. One path is the Flatten of the model input, which should result in (1, 4) not (4, 1)

2023-05-19T06:58:33.0585720Z ERROR: test_layer_normalization_2d_axis1_expanded_cpu 

2023-05-19T06:58:33.0593450Z onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Error compiling model compiler error:  Espresso exception: "Invalid blob shape": generic_elementwise_kernel: cannot broadcast:
2023-05-19T06:58:33.0593910Z 	(3, 4)
2023-05-19T06:58:33.0594210Z 	(4, 1)

I ran a couple of tools to a) add shape info to all the nodes, and b) run the 'basic' level optimizations like constant folding so we can see what nodes are around when the CoreML EP is creating the model.

```
python -m onnxruntime.tools.symbolic_shape_infer --input D:\src\github\ort\cmake\external\onnx\onnx\backend\test\data\node\test_layer_normalization_2d_axis1_expanded\model.onnx --output D:\src\github\ort\cmake\external\onnx\onnx\backend\test\data\node\test_layer_normalization_2d_axis1_expanded\model.ssi.onnx
python -m onnxruntime.tools.optimize_onnx_model --opt_level basic D:\src\github\ort\cmake\external\onnx\onnx\backend\test\data\node\test_layer_normalization_2d_axis1_expanded\model.ssi.onnx D:\src\github\ort\cmake\external\onnx\onnx\backend\test\data\node\test_layer_normalization_2d_axis1_expanded\model.ssi.basic.onnx
```

<img width="184" alt="image" src="https://github.com/microsoft/onnxruntime/assets/979079/88119032-f978-456c-acc5-c9cbe74c6025">

If I comment out Flatten in op_builder_factory.cc the tests pass.

skottmckay commented:
And this is the issue: 

https://github.com/microsoft/onnxruntime/blob/684e900e960cc5fbb2fcd1dd9e80ccb43a9fc310/onnxruntime/core/providers/coreml/builders/impl/flatten_op_builder.cc#L44

Rogue space in the attribute name so it wasn't reading the actual value from the node of 0.

I'll add a test and create a separate PR for that fix, but please test out your changes with just the fix ("axis " -> "axis") so we can check all CIs pass with that.

skottmckay commented:
Should be addressed by #16046.

skottmckay commented:
/azp run Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,MacOS CI Pipeline

skottmckay commented:
/azp run orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-python-checks-ci-pipeline,onnxruntime-binary-size-checks-ci-pipeline

skottmckay commented:
/azp run Windows ARM64 QNN CI Pipeline,Linux QNN CI Pipeline
